llvm/lib/CodeGen/MachineVerifier.cpp

```diff
@@ -1510,11 +1510,11 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
 
     LLT SrcTy = MRI->getType(MI->getOperand(NumDsts).getReg());
     if (DstTy.isVector()) {
-      // This case is the converse of G_CONCAT_VECTORS.
-      if (!SrcTy.isVector() ||
-          (SrcTy.getScalarType() != DstTy.getScalarType() &&
-           !SrcTy.isPointerVector()) ||
-          SrcTy.isScalableVector() != DstTy.isScalableVector() ||
+      // This case is the converse of G_CONCAT_VECTORS, but relaxed since
+      // G_UNMERGE_VALUES can handle src and dst vectors with different
+      // element sizes:
```

Comment:
It probably should not support this case. It's making the MIR much more complicated, forcing complexity into every piece of code that wants to do anything to a G_UNMERGE_VALUES. You should be required to unmerge by element, and insert explicit bitcasts

Reply:
Hm, so I could do this in `LegalizationArtifactCombiner::ArtifactValueFinder::tryCombineMergeLike`.
But wouldn't that *add* complexity in other code, such as the `tryCombineMergeLike`? Because I think this method is currently relying on unmerge to do the right thing:

https://github.com/llvm/llvm-project/blob/c13c04fdfe9b312160310eeeee785867627ffe2f/llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h#L998-L1021

Reply:
Ping.